File: 133.CloneGraph.py
class Solution(object):
    def cloneGraph(self, node):
        """
        :type node: UndirectedGraphNode
        :rtype: UndirectedGraphNode
        """
        queue = []
        dic = {}
        if not node:
            return None
        queue.append(node)
        start = 0
        while start < len(queue):
            cur = queue[start]
            neighbors = cur.neighbors
            if cur not in dic:
                dic[cur] = UndirectedGraphNode(cur.label)
                for item in neighbors:
                    queue.append(item)
            start+=1
        for item in dic:
            neighbors = item.neighbors
            for neighbor in neighbors:
                dic[item].neighbors.append(dic[neighbor])
        return dic[node]

File: 133.CloneGraph.java
/**
 * Definition for undirected graph.
 * class UndirectedGraphNode {
 *     int label;
 *     List<UndirectedGraphNode> neighbors;
 *     UndirectedGraphNode(int x) { label = x; neighbors = new ArrayList<UndirectedGraphNode>(); }
 * };
 */
public class Solution {
    public UndirectedGraphNode cloneGraph(UndirectedGraphNode node) {
        if(node==null){
            return null;
        }
        HashMap<UndirectedGraphNode, UndirectedGraphNode> map = new HashMap<>();
        return copy(node, map);
    }
    private UndirectedGraphNode copy(UndirectedGraphNode node, HashMap<UndirectedGraphNode, UndirectedGraphNode> map){
        UndirectedGraphNode newnode = new UndirectedGraphNode(node.label);
        map.put(node, newnode);
        List<UndirectedGraphNode> neighbors = node.neighbors;
        for(UndirectedGraphNode nei:neighbors){
            if(map.containsKey(nei)){
                newnode.neighbors.add(map.get(nei));
            }else{
                newnode.neighbors.add(copy(nei, map));
            }
        }
        return newnode;
    }
}
